source/common/config/grpc_mux_impl.cc

```diff
@@ -190,17 +192,46 @@ void GrpcMuxImpl::onDiscoveryResponse(
     std::vector<DecodedResourceRef> all_resource_refs;
     OpaqueResourceDecoder& resource_decoder =
         api_state_[type_url].watches_.front()->resource_decoder_;
+
+    auto ttl_expiry_callback = [this, type_url](const auto& expired) {
+      absl::flat_hash_set<std::string> all_expired;
+      all_expired.insert(expired.begin(), expired.end());
+
+      for (auto watch : api_state_[type_url].watches_) {
+        std::vector<std::string> found_resources;
+
+        for (const auto& resource : expired) {
+          if (all_expired.find(resource) != all_expired.end()) {
+            found_resources.push_back(resource);
+          }
+        }
+
+        watch->callbacks_.onConfigExpired(found_resources);
+      }
+    };
+
+    api_state_[type_url].ensureTtlManagerCreated(dispatcher_, ttl_expiry_callback);
+    const auto scoped_ttl_update = api_state_[type_url].ttl_->scopedTtlUpdate();
     for (const auto& resource : message->resources()) {
-      if (message->type_url() != resource.type_url()) {
+      // TODO(snowp): Check the underlying type when the resource is a Resource.
+      if (!resource.Is<envoy::service::discovery::v3::Resource>() &&
```

Comment:
FWIW, this is another place that having two GrpcMuxImpls is hurting us, CC @wgallagher @dmitri-d